hivemind/utils/logging.py

```diff
@@ -60,23 +68,74 @@ def format(self, record: logging.LogRecord) -> str:
         return super().format(record)
 
 
-def get_logger(module_name: str) -> logging.Logger:
-    # trim package name
-    name_without_prefix = ".".join(module_name.split(".")[1:])
+_PACKAGE_NAME = __name__.split(".")[0]
 
-    logging.addLevelName(logging.WARNING, "WARN")
-    formatter = CustomFormatter(
-        fmt="{asctime}.{msecs:03.0f} [{bold}{levelcolor}{levelname}{reset}] [{bold}{caller}{reset}] {message}",
-        style="{",
-        datefmt="%b %d %H:%M:%S",
-    )
-    handler = logging.StreamHandler()
-    handler.setFormatter(formatter)
-    logger = logging.getLogger(name_without_prefix)
-    logger.setLevel(loglevel)
-    logger.addHandler(handler)
+_init_lock = threading.RLock()
+_default_handler = None
+
+
+def _initialize_if_necessary():
+    global _current_mode, _default_handler
+
+    with _init_lock:
+        if _default_handler is not None:
+            return
+
+        formatter = CustomFormatter(
+            fmt="{asctime}.{msecs:03.0f} [{bold}{levelcolor}{levelname}{reset}] [{bold}{caller}{reset}] {message}",
+            style="{",
+            datefmt="%b %d %H:%M:%S",
+        )
+        _default_handler = logging.StreamHandler()
+        _default_handler.setFormatter(formatter)
+
+        _current_mode = StyleMode.NOWHERE  # Corresponds to the initial logger state
+        use_hivemind_log_style(StyleMode.AMONG_HIVEMIND)  # Overriding it to the desired default
+
+
+def get_logger(name: Optional[str] = None) -> logging.Logger:
+    _initialize_if_necessary()
+    return logging.getLogger(name)
+
+
+def _enable_default_handler(name: str) -> None:
+    logger = get_logger(name)
+    logger.addHandler(_default_handler)
     logger.propagate = False
-    return logger
+    logger.setLevel(loglevel)
+
+
+def _disable_default_handler(name: str) -> None:
+    logger = get_logger(name)
+    logger.removeHandler(_default_handler)
+    logger.propagate = True
+    logger.setLevel(logging.NOTSET)
+
+
+class StyleMode(Enum):
+    NOWHERE = 0
+    AMONG_HIVEMIND = 1
+    EVERYWHERE = 2
+
+
+def use_hivemind_log_style(where: Union[StyleMode, str]) -> None:
+    global _current_mode
+
+    if isinstance(where, str):
+        # We allow `where` to be a string, so a developer does not have to import the enum for one usage
+        where = StyleMode[where.upper()]
+
+    if _current_mode == StyleMode.AMONG_HIVEMIND:
+        _disable_default_handler(_PACKAGE_NAME)
+    elif _current_mode == StyleMode.EVERYWHERE:
+        _disable_default_handler(None)
```

Comment:
I'm not sure if this is going to be a frequent enough use case, aside from our examples. Hivemind is first and foremost a library, and from my understanding, no other library implements this format hijacking mechanism: thus, I am not sure that this feature is actually required and expected to be present in a library about distributed DL. Probably, the reason for this is that it's usually expected that logs from the given library are consistent in their format in all applications for ease of support and issue reporting, and thus one has no direct incentive to implicitly adopt the logging format of other library

Reply:
This isn't a blocker, I'm just afraid that adding unnecessary/rarely used features should not be our focus regarding logging